### PR TITLE
[Merged by Bors] - feat: volume of a complex ball

### PR DESCRIPTION
**Pull request by xroblot**
We prove the formula for the area of a disc
```lean
theorem volume_ball (x : EuclideanSpace ℝ (Fin 2)) (r : ℝ) :
    volume (Metric.ball x r) = NNReal.pi * (ENNReal.ofReal r) ^ 2 
```
and deduce from this, the volume of complex balls 
```lean
theorem volume_ball (a : ℂ) (r : ℝ) :  
    volume (Metric.ball a r) = NNReal.pi * (ENNReal.ofReal r) ^ 2
```

Co-authored-by: James Arthur
Co-authored-by: Benjamin Davidson
Co-authored-by: Andrew Souther

---

- [x] depends on: #6832 
- [x] depends on: #6905 
- [x] depends on: #7037 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jsm28** [comment]
Can `Archive.Wiedijk100Theorems.AreaOfACircle` be refactored in terms of this (or alternatively the `100.yaml` entry repointed to this result, with something done about the doc string discussing how there should be a more general result for the n-ball)? I don't think we should need to have two completely separate proofs of essentially the same result.

**Comment by xroblot** [comment]
> Can `Archive.Wiedijk100Theorems.AreaOfACircle` be refactored in terms of this (or alternatively the `100.yaml` entry repointed to this result, with something done about the doc string discussing how there should be a more general result for the n-ball)? I don't think we should need to have two completely separate proofs of essentially the same result.

As the matter of fact, I took some ideas from `AreaOfACircle` for this PR (and I am planning to add the authors of this file as co-authors of this PR when I write down the description), the main difference being the way the integral of $\sqrt {1 - x ^2 }$ is computed. I did not explicitly state the result about the area of the disc since I was not sure where to put it, but now that I rethink about it, it could belong in the`EuclideanSpace` namespace (with a doc string about the $n$-dimensional ball). 

In any case, I think it is fine to keep `AreaOfACircle` with the same proof since it is in `Archive` and was historically the first one to solve the corresponding `100.yaml` problem.

**Comment by hrmacbeth** [comment]
> with something done about the doc string discussing how there should be a more general result for the n-ball

By the way, @fpvandoorn and I recently proved the n-ball result.  We will PR it soon (I hope) but had not discussed where it should go.

**Comment by xroblot** [comment]
> > with something done about the doc string discussing how there should be a more general result for the n-ball
> 
> By the way, @fpvandoorn and I recently proved the n-ball result. We will PR it soon (I hope) but had not discussed where it should go.

Great! I'll will wait until you PR your result with Floris and then use it to prove the volume of the complex ball. 

Do you think #6905 is still worth keeping?

**Comment by hrmacbeth** [comment]
Actually, I think there's no reason to kill this PR -- it can join mathlib now and when ours is ready we can generalize it.

**Comment by xroblot** [comment]
Ok, in that case, I think it would be better to also have the result on the area of the disc in this PR and the question of where it belongs is still open...

**Comment by xroblot** [comment]
So we do have [Euclidean.ball](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/InnerProductSpace/EuclideanDist.html#Euclidean.ball)

**Comment by eric-wieser** [comment]
I think `Euclidean.ball` is intended only as an implementation detail elsewhere

**Comment by xroblot** [comment]
> I think `Euclidean.ball` is intended only as an implementation detail elsewhere

So something like
```lean
example : volume (Euclidean.ball (0 : Fin 2 → ℝ) 1) = NNReal.pi 
```
is not the right way to state the result?

**Comment by eric-wieser** [comment]
> In any case, I think it is fine to keep `AreaOfACircle` with the same proof since it is in `Archive` and was historically the first one to solve the corresponding `100.yaml` problem.

I think we should replace the proof anyway, though perhaps leave the statement as is for now. If people want to see the historical solution, they can always look at the git history for that file. If you don't agree, it's perhaps worth starting a Zulip thread to see what the wider opinion on this is.

My view is that `Archive` should be demonstrations of the best way to do things with *current* mathlib, not just a historical leaderboard.

**Comment by xroblot** [comment]
> By the way, @fpvandoorn and I recently proved the n-ball result. We will PR it soon (I hope) but had not discussed where it should go.

@hrmacbeth can you please share the statement that you proved with Floris. I am still not sure of the best way to state the dimension 2 version...

**Comment by fpvandoorn** [comment]
Our result is here:
https://github.com/leanprover-community/mathlib4/blob/e27595ad3a9d1b8b77138c3164e0bed8aa1703ef/Mathlib/MeasureTheory/Measure/Lebesgue/VolumeBall.lean#L212-L213

the definition of `B` is here:
https://github.com/leanprover-community/mathlib4/blob/e27595ad3a9d1b8b77138c3164e0bed8aa1703ef/Mathlib/MeasureTheory/Measure/Lebesgue/VolumeBall.lean#L114

**Comment by eric-wieser** [comment]
> > I think `Euclidean.ball` is intended only as an implementation detail elsewhere
> 
> So something like
> 
> ```lean
> example : volume (Euclidean.ball (0 : Fin 2 → ℝ) 1) = NNReal.pi 
> ```
> 
> is not the right way to state the result?

Correct, the right way is with `volume (ball (0 : EuclideanSpace (Fin 2) ℝ) 1)` on the LHS.

File: Mathlib/Analysis/NormedSpace/PiLp.lean [hunk 1]
```diff
@@ -81,6 +81,9 @@ abbrev PiLp (p : ℝ≥0∞) {ι : Type*} (α : ι → Type*) : Type _ :=
 instance (p : ℝ≥0∞) {ι : Type*} (α : ι → Type*) [∀ i, Inhabited (α i)] : Inhabited (PiLp p α) :=
   ⟨fun _ => default⟩
 
+instance (p : ℝ≥0∞) {ι : Type*} [Inhabited ι] (α : ι → Type*) [∀ (i : ι), Nonempty (α i)]
+    [Nontrivial (α default)] : Nontrivial (PiLp p α) := Pi.nontrivial
```
Comment (eric-wieser) [inline Mathlib/Analysis/NormedSpace/PiLp.lean:85]: I think this should be an instance in `WithLp.lean`, that says "if `V` is nontrivial then so is `WithLp p V`"

Comment (xroblot) [reply Mathlib/Analysis/NormedSpace/PiLp.lean]: Yes, that makes sense. Here it comes.

File: Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean [hunk 2]
```diff
@@ -31,11 +32,26 @@ def measurableEquivPi : ℂ ≃ᵐ (Fin 2 → ℝ) :=
   basisOneI.equivFun.toContinuousLinearEquiv.toHomeomorph.toMeasurableEquiv
 #align complex.measurable_equiv_pi Complex.measurableEquivPi
 
+@[simp]
+theorem measurableEquivPi_apply (a : ℂ) :
+    measurableEquivPi a = Fin.cons a.re (Fin.cons a.im finZeroElim) := rfl
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean:37]: Is `![]` notation not available here?

Comment (xroblot) [reply Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean:37]: Indeed, it works here and makes things easier to read. Thanks

**Review by eric-wieser — COMMENTED** [review]
Do you think it makes sense to add the `ProdLp` case now too (refactoring the existing proof), or leave that to a future PR?

**Comment by xroblot** [comment]
> Do you think it makes sense to add the `ProdLp` case now too (refactoring the existing proof), or leave that to a future PR?

I think that will be for a future PR. The proof for this case is completely different than the one in this PR and, in any case, I think the more general proof in the future PR by @hrmacbeth and @fpvandoorn should be easier to adapt to the case `Lp`.

File: Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean [hunk 3]
```diff
@@ -49,4 +65,25 @@ theorem volume_preserving_equiv_real_prod : MeasurePreserving measurableEquivRea
...
+    measurableSet_ball]
+  refine congrArg _ (Set.ext fun _ => ?_)
+  simp_rw [← MeasurableEquiv.coe_toEquiv_symm, Set.mem_preimage, MeasurableEquiv.coe_toEquiv_symm,
+    measurableEquivPi_symm_apply, mem_ball_zero_iff, norm_eq_abs, abs_def, normSq_add_mul_I,
+    EuclideanSpace.coe_measurableEquiv_symm, EuclideanSpace.norm_eq, WithLp.equiv_symm_pi_apply,
+    Fin.sum_univ_two, Real.norm_eq_abs, _root_.sq_abs]
+
+@[simp]
+theorem volume_closedBall (a : ℂ) (r : ℝ) :
+     volume (Metric.closedBall a r) = NNReal.pi * (ENNReal.ofReal r) ^ 2 := by
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean:86]:
```suggestion
     volume (Metric.closedBall a r) = NNReal.pi * ENNReal.ofReal r ^ 2 := by
```

File: Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean [hunk 4]
```diff
@@ -49,4 +65,25 @@ theorem volume_preserving_equiv_real_prod : MeasurePreserving measurableEquivRea
   (volume_preserving_finTwoArrow ℝ).comp volume_preserving_equiv_pi
 #align complex.volume_preserving_equiv_real_prod Complex.volume_preserving_equiv_real_prod
 
+-- See: https://github.com/leanprover/lean4/issues/2220
+local macro_rules | `($x ^ $y)   => `(HPow.hPow $x $y)
+
+@[simp]
+theorem volume_ball (a : ℂ) (r : ℝ) :
+    volume (Metric.ball a r) = NNReal.pi * (ENNReal.ofReal r) ^ 2 := by
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Measure/Lebesgue/Complex.lean:73]:
```suggestion
    volume (Metric.ball a r) = NNReal.pi * ENNReal.ofReal r ^ 2 := by
```

File: Mathlib/MeasureTheory/Measure/Haar/InnerProductSpace.lean [hunk 5]
```diff
@@ -100,3 +103,49 @@ theorem PiLp.volume_preserving_equiv_symm : MeasurePreserving (WithLp.equiv 2 (
   (EuclideanSpace.volume_preserving_measurableEquiv ι).symm
 
 end PiLp
+
+namespace EuclideanSpace
+
+open BigOperators ENNReal
+
+-- See: https://github.com/leanprover/lean4/issues/2220
+local macro_rules | `($x ^ $y)   => `(HPow.hPow $x $y)
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Measure/Haar/InnerProductSpace.lean:112]: Same here: please copy the standard hack + comment, and put it at the top of the file.

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks! sorry for forgetting about this.

**Comment by xroblot** [comment]
> bors d+
> 
> Thanks! sorry for forgetting about this.

Thanks. In fact, I have a new PR coming soon with the computation of the volume of the unit ball in all dimensions, for all Lp norms over the real and over the complex using #7693. I just need to do some clean up and finish proving one or two easy lemmas (famous last words).

**Pull request closed without merging**